tempesta_fw/http.c

```diff
@@ -414,30 +417,45 @@ tfw_http_conn_init(TfwConnection *conn)
 }
 
 /*
- * Connection with a peer is released.
+ * Destroy server connection's resources.
  *
- * For server connections requests that were sent to that server are kept
+ * For server connections, requests that were sent to that server are kept
  * in the queue until a paired response comes. That will never happen now.
  * For each request that has been unanswered send an error response, then
  * delete the request and drop the connection with the client if required.
- *
- * Called when a connection is released. There are no users at that time,
- * so locks are not needed.
  */
 static void
-tfw_http_conn_release(TfwConnection *conn)
+tfw_http_conn_srv_release(TfwConnection *srv_conn)
 {
-	TfwMsg *msg, *tmp;
-
-	list_for_each_entry_safe(msg, tmp, &conn->msg_queue, msg_list) {
-		BUG_ON(((TfwHttpMsg *)msg)->conn
-			&& (((TfwHttpMsg *)msg)->conn == conn));
-		list_del(&msg->msg_list);
-		tfw_http_send_404((TfwHttpMsg *)msg);
-		tfw_http_conn_cli_dropfree((TfwHttpMsg *)msg);
+	TfwHttpMsg *hmreq, *tmp;
+	struct list_head zap_queue;
+
+	TFW_DBG3("%s: conn = %p\n", __func__, srv_conn);
+	BUG_ON(!(TFW_CONN_TYPE(srv_conn) & Conn_Srv));
+
+	INIT_LIST_HEAD(&zap_queue);
+
+	spin_lock(&srv_conn->msg_qlock);
+	list_splice_tail_init(&srv_conn->msg_queue, &zap_queue);
+	spin_unlock(&srv_conn->msg_qlock);
+
+	list_for_each_entry_safe(hmreq, tmp, &zap_queue, msg.fwd_list) {
+		BUG_ON((hmreq)->conn && ((hmreq)->conn == srv_conn));
+		list_del_init(&hmreq->msg.fwd_list);
+		tfw_http_send_404(hmreq);
 		TFW_INC_STAT_BH(clnt.msgs_otherr);
 	}
-	INIT_LIST_HEAD(&conn->msg_queue);
+}
+
+/*
+ * Connection with a peer is released.
+ * Destroy connection's remaining resources.
+ */
+static void
+tfw_http_conn_release(TfwConnection *conn)
+{
+	if (TFW_CONN_TYPE(conn) & Conn_Srv)
+		tfw_http_conn_srv_release(conn);
```

Comment:
Well, if we don't need to do anything for client connections, then lets rework `tfw_connection_release()` as well as all the callback calls - there is no sense to call callback just to check that we don't need to do anything.